### ISSUE TITLE
New rule suggestion: literal conventions mixed

### ISSUE THREAD

**Issue by jrencz** [eslint/eslint#1614]
With `space-in-brackets` enabled I had an error like this:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

I got noticed by `space-in-brackets`

```
error  There should be no space after '['  space-in-brackets
```

Since there are certainly spaces after the opening bracket this rule fails correctly, but it's not the most clear message possible.

The same occurs when array is written in consistently, but the object literal is not:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.',
].join('\n'), {
    foo: 'foo',
    bar: 'bar'});
```

```
error  There should be no space after '{'  space-in-brackets
```
However using new lines inside the literal without spaces after left bracket/before right brakcet is ok for this rule:
 
```
_.template(['Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

The error which caused the problem is the mix of conventions: first element of the array is in the new line, but there is no new line after the last. The correct form of the snippet (with both literals expanded) will be:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.',
].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

I suggest adding new rule rule which would warn if conventions are mixed.
Already existing rules do not seem to cover the case: I can either force or disable spaces in brackets but there is no rule which explicitly addresses the case checking whether the literal (both array and object i suppose) is written consistently.

My eslint configuration related to stylistic issues:
```
"brace-style": 2,
"camelcase": 0,
"new-cap": 0,
"no-lonely-if": 2,
"no-underscore-dangle": 0,
"space-after-keywords": 2,
"space-in-brackets": [2, "never"],
"space-infix-ops": 2,
"quotes": [2, "single"],
```

**Comment by jrencz** [comment]
I use eslint via grunt-eslint `2.1.0` and it has `^0.10.0` with `0.10.2` currently installed in my setup.
I checked on v0.11.0 as well without grunt and linting the following snippet:

```
_.template([
    'Foo: ${ foo }',
    'Bar: `${ bar }`.'].join('\n'), {
    foo: 'foo',
    bar: 'bar',
});
```

results in the same error:

```
  1:11  error  There should be no space after '['  space-in-brackets
```

**Comment by nzakas** [comment]
OK, that's definitely a bug. We should be allowing newlines there.

**Comment by btmills** [comment]
@nzakas to confirm my understanding of the discussion here, the rule is currently designed to enforce spacing unless both the starting and ending brackets are on separate lines, which is a bug. Instead, when only one bracket is on a separate line, it should enforce spacing only for the other bracket. Is that correct?

```js
// Enforced currently and with the proposed change
var arr = [1, 2, 3, 4];

// Not enforced currently, nor with the proposed change
var arr = [
    1, 2, 3, 4
];

// Beginning and end enforced currently
var arr = [1, 2, // Only beginning enforced with the proposed change
    3, 4
]; // Ending not enforced with the proposed change
```

The following four test cases are pulled directly from `tests/lib/rules/space-in-brackets.js`. According to the rule's original design, there should be one error each. Am I reading that this behavior is incorrect, and all four should pass validation?

```js
// "space-in-brackets": [2, "never"]
var obj = {foo: bar,
baz: qux
};

// "space-in-brackets": [2, "never"]
var obj = {
foo: bar,
baz: qux};

// "space-in-brackets": [2, "never"]
var arr = [1,
2,
3,
4
];

// "space-in-brackets": [2, "never"]
var arr = [
1,
2,
3,
4];
```

If my interpretation is correct, the change seems strange to me. In fact, I find the spacing in the example given to be confusing: the elements of an array expression and object literal are at the same indentation level on adjacent lines, and the divide between the two gets buried in the third line. I think the given snippet could be improved with either of the following changes, both of which validate without altering this rule:

```js
_.template(
	[
		'Foo: ${ foo }',
		'Bar: `${ bar }`.'
	].join('\n'),
	{ foo: 'foo', bar: 'bar' }
);

_.template([
	'Foo: ${ foo }',
	'Bar: `${ bar }`.'
].join('\n'), {
	foo: 'foo',
	bar: 'bar',
});
```

Am I misunderstanding? Or is my eye alone in finding those four test cases kind of ugly?

**Comment by nzakas** [comment]
I don't like judging other people's coding style - I want ESLint to be able to accommodate and style based on obvious rules.

Your understanding is correct, but I'm not sure about the four test cases. They seem like they shouldn't have errors. Is it flagging the brace that isn't on the same line? If so, that means the rule also enforces having braces on the same line, which seems wrong.

**Comment by btmills** [comment]
Very well. As for the four test cases, they are verifying the current behavior, which requires that both brackets are on the same line as the adjacent element, in which case spacing is enforced, or both brackets are on separate lines, and spacing is unenforced. These four started failing when I changed the logic to check beginning and ending spacing separately.

**Comment by btmills** [comment]
Working on this.
